### PR TITLE
Adds Polychromic Clothes

### PR DESCRIPTION
**Pull request by Michiyamenotehifunana**
:cl: Toriate
add: Added Polychromic clothes
tweak: Made mixed wardrobes spawn polychromic clothes
imageadd: added greyscaled sprites for polychromic clothes
/:cl:

dreamseeker_2017-11-19_22-53-16

Polychromic clothes are recolourable by alt clicking them. They can have up to 3 recolourable overlays, and have a base sprite containing unrecolourable elements.

This PR introduces a framework for polychromic clothes, anyone can make one by providing up to 4 sprites: A base one, and up to 3 recolorable overlays for it, and then copypasting some code. This PR includes 5 example clothes, all spawning in mixed wardrobes.

### CODE REVIEW

**Comment by Kevak** [comment]
Looks like we need to get the tagging system working properly. Darnit.

I'll add WIP. Ping me when its ready. I know yer fiddling with it.

**Comment by Kevak** [comment]
Once I figure out how to add tags.

**Comment by Michiyamenotehifunana** [comment]
Technically it's actually ready for merging, but I might as well polish it up. I just need to do list shenanigans that only affects the framework and not any of the example clothes added.

**Comment by Kevak** [comment]
Alrighty. Ping when yer ready for review.

**Comment by Michiyamenotehifunana** [comment]
@Kevak beep boop meow meow

**Comment by ActualRory** [comment]
How many times can you change the colour of the clothing?
Is it an initial setting, or can you change at will?

**Comment by Michiyamenotehifunana** [comment]
You can change it at will, in real time.

**Comment by Kevak** [comment]
Code all looks good but ya need to comment yer code. A big issue in a lotta codebases is code that isn't fully commented and explained. Please do so. Explain what each proc does in //comments

**Comment by Partheo** [comment]
Doesn't actually modify/improve any current code, leave the "Code improvement" tag for that.

File: haven/code/polychromic_clothes.dm [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+/obj/item/clothing/under/polychromic/update_icon()
+	..()
+	cut_overlays()
+
+	if(hasprimary)
+		var/mutable_appearance/primary_overlay = mutable_appearance('haven/icons/polyclothes/item/uniform.dmi', "[item_color]-primary")
+		primary_overlay.color = primary_color
+		add_overlay(primary_overlay)
+
```
Comment (Partheo) [inline haven/code/polychromic_clothes.dm]: Stacked IF statements shouldn't have whitespace between them.

File: haven/code/polychromic_clothes.dm [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+			secondary_worn.color = secondary_color
+			. += secondary_worn
+		if(hastertiary)
+			var/mutable_appearance/tertiary_worn = mutable_appearance('haven/icons/polyclothes/mob/uniform.dmi', "[item_color]-tertiary")
+			tertiary_worn.color = tertiary_color
+			. += tertiary_worn
+
+/obj/item/clothing/under/polychromic/examine(mob/user)
+	..()
+	to_chat(user, "<span class='notice'>Alt-click to recolor it.</span>")
```
Comment (Partheo) [inline haven/code/polychromic_clothes.dm]: Muh immershuns

**Comment by Kevak** [comment]
It adds the framework for recolorable stuff. So technically it is an improvement imo.

File: haven/code/polychromic_clothes.dm [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	item_color = "polyjump"
+	item_state = "rainbow"
+	hasprimary = TRUE
+	hassecondary = TRUE
+	hastertiary = TRUE
+	primary_color = "#FFFFFF" //RGB in hexcode
+	secondary_color = "#808080"
+	tertiary_color = "#FF3535"
+
+//replaces the jumpsuit contents of the mixed wardrobe
```
Comment (Partheo) [inline haven/code/polychromic_clothes.dm]: How do you know that this overwrites the TG proc?

**Review by Partheo — CHANGES_REQUESTED** [review]
Please leave the code for the clothes in /haven/code/modules/clothing/under/polychromatic_clothes.dm, and the code for the wardrobe in /haven/code/game/objects/structures/crates_lockers/closets/wardrome.dm. I know it's a pain, but that's where the code should go were it not in the modular folder. If you want, I can do that myself.

Also, since you're overwriting a TG proc, please update haven/changes.md.

File: haven/code/polychromic_clothes.dm [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	if(hastertiary)
+		var/mutable_appearance/tertiary_overlay = mutable_appearance('haven/icons/polyclothes/item/uniform.dmi', "[item_color]-tertiary")
+		tertiary_overlay.color = tertiary_color
+		add_overlay(tertiary_overlay)
+
+/obj/item/clothing/under/polychromic/AltClick(mob/living/user)
+	if(user.incapacitated() || !istype(user))
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (Partheo) [inline haven/code/polychromic_clothes.dm]: You should check if they're in range first.

File: haven/code/polychromic_clothes.dm [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+		return
+	if(user.incapacitated() || !istype(user) || !in_range(src, user))
```
Comment (Partheo) [inline haven/code/polychromic_clothes.dm]: This check is redundant, since you've already checked for these just before.

**Review by Partheo — CHANGES_REQUESTED** [review]

**Comment by Michiyamenotehifunana** [comment]
_Should_ be done properly now.

**Review by Partheo — APPROVED** [review]

**Comment by Partheo** [comment]
Looks good. I'll test this today unless Neri wants to.

**Comment by Kevak** [comment]
Go ahead

**Pull request merged**